File: leetcode-in-python/solution/35_search_insert_position.py
from typing import List


class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums)
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if nums[mid] < target: # 绝对小于，右半边
                lo = mid + 1
            else:
                hi = mid
        return hi


import unittest


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.input = [
            ([1,3,4,6], 7),
            ([4,5,5,5,], 5),
        ]
        self.answer = [
            4,
            1,
        ]
        self.s = Solution()

    def test_solution(self):
        for i, (item, target) in enumerate(self.input):
            ans = self.s.searchInsert(item, target)
            self.assertEqual(ans , self.answer[i],ans )


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_35_search_insert_position.rs
pub struct Solution {}

impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        let (mut lo, mut hi) = (0, nums.len());
        while lo < hi {
            let mid = lo + (hi - lo) / 2;
            if nums[mid] < target {
                lo = mid + 1
            } else {
                hi = mid
            }
        }
        hi as i32
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (vec![1, 3, 4, 6], 0, 0, ),
            (vec![1, 3, 4, 6], 2, 1, ),
            (vec![1, 3, 4, 6], 5, 3, ),
            (vec![1, 3, 4, 6], 7, 4, ),
            (vec![4, 5, 5, 5], 5, 1, ),
            (vec![5, 5, 5, 5], 5, 0, ),
        ];


        for (nums, target, answer) in test_case {
            let ans = Solution::search_insert(nums, target);
            assert_eq!(ans, answer);
        }
    }
}
